### PR TITLE
Move scheduler extender API V1 to staging k8s.io/kube-scheduler

### LINKED ISSUE #88536: Move Scheduler Extender API to staging
The types for sending/receiving requests between the scheduler and the extenders currently exist in pkg/scheduler/apis/v1/extender, and so extender developers have to import the kubernetes package to access those types, which brings with it many unneeded dependencies.

To be consistent with the way we handle component config, and simplify importing those types, we should have the extender types moved to staging/src/k8s.io/kube-scheduler.

/assign @damemi 
/sig scheduling

/cc @alculquicondor @Huang-Wei

### PR DESCRIPTION
**Pull request by damemi**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This moves the Extender api for the scheduler, previously under `pkg/scheduler/apis/extender/v1` to the `k8s.io/kube-scheduler` staging repo

**Which issue(s) this PR fixes**: 
Fixes #88536 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Scheduler Extender API is now located under k8s.io/kube-scheduler/extender
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by ahg-g** [comment]
I think we need to define an internal type that the versioned type gets converted to.

**Comment by liggitt** [comment]
There are quite a few unusual things going on with this types:

1. No json tags defined (so all fields serialize capitalized)
2. No internal type or conversions defined
3. No TypeMeta fields in the objects (so no apiVersion/kind identification is possible)

I'd like to understand more about how these are used/documented today

**Comment by ahg-g** [comment]
I don't know the history of those types, and why they are not properly setup, this API seems to have been around for sometime. 

This API is used to define request/reply types for the scheduler to call out to external services (called Extenders) to make decisions when scheduling a pod (filtering, scoring and preemption).

The only documentation I could find was this: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/scheduler_extender.md

**Comment by alculquicondor** [comment]
I think we can all agree that this API is not well defined.

However, we can not fix it easily. So I think we have to add some api-violation exceptions and then release a v2 next cycle.

**Comment by liggitt** [comment]
Add an OWNERS file to the extenders dir:

```
# See the OWNERS docs at https://go.k8s.io/owners

# Disable inheritance as this is an api owners file
options:
  no_parent_owners: true
approvers:
- api-approvers
reviewers:
- api-reviewers
- sig-scheduling
labels:
- kind/api-change
- sig/scheduling
```

**Comment by liggitt** [comment]
> I think we can all agree that this API is not well defined.
> 
> However, we can not fix it easily. So I think we have to add some api-violation exceptions and then release a v2 next cycle.

Are there any tests that would catch attempts to "fix" these types by adding lowercased json tags (since that would break compatibility with existing extenders)?

**Comment by liggitt** [comment]
Locating the types in an exported package makes sense. Make sure there's an OWNERS file covering the directory that routes to API reviewers/approvers, and add a test that ensures the following work, then lgtm:
* populating a struct, then encoding, produces upper-case field names
* upper-case field names can be decoded into the struct
* lower-case field names can be decoded into the struct

**Comment by damemi** [comment]
@liggitt I added the test and OWNERS you asked for. Is there a way to make this kind of testing table driven? Since it's all different types I couldn't think of an easy one besides manually roundtripping each type

**Comment by liggitt** [comment]
yes, can make a test like this:
```go

import (
	"encoding/json"
	"reflect"
	"strings"
	"testing"

	"github.com/google/go-cmp/cmp"

	corev1 "k8s.io/api/core/v1"
	v1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/types"
)

func TestCompatibility(t *testing.T) {
	testcases := []struct {
		emptyObj   interface{}
		obj        interface{}
		expectJSON string
	}{
		{
			emptyObj: &ExtenderPreemptionResult{},
			obj: &ExtenderPreemptionResult{
				NodeNameToMetaVictims: map[string]*MetaVictims{"foo": &MetaVictims{Pods: []*MetaPod{{UID: "myuid"}}, NumPDBViolations: 1}},
			},
			expectJSON: `{"NodeNameToMetaVictims":{"foo":{"Pods":[{"UID":"myuid"}],"NumPDBViolations":1}}}`,
		},
		{
			emptyObj: &ExtenderPreemptionArgs{},
			obj: &ExtenderPreemptionArgs{
				Pod:                   &corev1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "podname"}},
				NodeNameToVictims:     map[string]*Victims{"foo": &Victims{Pods: []*v1.Pod{&corev1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "podname"}}}, NumPDBViolations: 1}},
				NodeNameToMetaVictims: map[string]*MetaVictims{"foo": &MetaVictims{Pods: []*MetaPod{{UID: "myuid"}}, NumPDBViolations: 1}},
			},
			expectJSON: `{"Pod":{"metadata":{"name":"podname","creationTimestamp":null},"spec":{"containers":null},"status":{}},"NodeNameToVictims":{"foo":{"Pods":[{"metadata":{"name":"podname","creationTimestamp":null},"spec":{"containers":null},"status":{}}],"NumPDBViolations":1}},"NodeNameToMetaVictims":{"foo":{"Pods":[{"UID":"myuid"}],"NumPDBViolations":1}}}`,
		},
		{
			emptyObj: &ExtenderArgs{},
			obj: &ExtenderArgs{
				Pod:       &corev1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "podname"}},
				Nodes:     &corev1.NodeList{Items: []corev1.Node{corev1.Node{ObjectMeta: metav1.ObjectMeta{Name: "nodename"}}}},
				NodeNames: &[]string{"node1"},
			},
			expectJSON: `{"Pod":{"metadata":{"name":"podname","creationTimestamp":null},"spec":{"containers":null},"status":{}},"Nodes":{"metadata":{},"items":[{"metadata":{"name":"nodename","creationTimestamp":null},"spec":{},"status":{"daemonEndpoints":{"kubeletEndpoint":{"Port":0}},"nodeInfo":{"machineID":"","systemUUID":"","bootID":"","kernelVersion":"","osImage":"","containerRuntimeVersion":"","kubeletVersion":"","kubeProxyVersion":"","operatingSystem":"","architecture":""}}}]},"NodeNames":["node1"]}`,
		},
		{
			emptyObj: &ExtenderFilterResult{},
			obj: &ExtenderFilterResult{
				Nodes:       &corev1.NodeList{Items: []corev1.Node{corev1.Node{ObjectMeta: metav1.ObjectMeta{Name: "nodename"}}}},
				NodeNames:   &[]string{"node1"},
				FailedNodes: FailedNodesMap{"foo": "bar"},
				Error:       "myerror",
			},
			expectJSON: `{"Nodes":{"metadata":{},"items":[{"metadata":{"name":"nodename","creationTimestamp":null},"spec":{},"status":{"daemonEndpoints":{"kubeletEndpoint":{"Port":0}},"nodeInfo":{"machineID":"","systemUUID":"","bootID":"","kernelVersion":"","osImage":"","containerRuntimeVersion":"","kubeletVersion":"","kubeProxyVersion":"","operatingSystem":"","architecture":""}}}]},"NodeNames":["node1"],"FailedNodes":{"foo":"bar"},"Error":"myerror"}`,
		},
		{
			emptyObj: &ExtenderBindingArgs{},
			obj: &ExtenderBindingArgs{
				PodName:      "mypodname",
				PodNamespace: "mypodnamespace",
				PodUID:       types.UID("mypoduid"),
				Node:         "mynode",
			},
			expectJSON: `{"PodName":"mypodname","PodNamespace":"mypodnamespace","PodUID":"mypoduid","Node":"mynode"}`,
		},
		{
			emptyObj:   &ExtenderBindingResult{},
			obj:        &ExtenderBindingResult{Error: "myerror"},
			expectJSON: `{"Error":"myerror"}`,
		},
		{
			emptyObj:   &HostPriority{},
			obj:        &HostPriority{Host: "myhost", Score: 1},
			expectJSON: `{"Host":"myhost","Score":1}`,
		},
	}

	for _, tc := range testcases {
		t.Run(reflect.TypeOf(tc.obj).String(), func(t *testing.T) {
			data, err := json.Marshal(tc.obj)
			if err != nil {
				t.Fatal(err)
			}
			if string(data) != tc.expectJSON {
				t.Fatalf("expected %s, got %s", tc.expectJSON, string(data))
			}
			if err := json.Unmarshal([]byte(strings.ToLower(string(data))), tc.emptyObj); err != nil {
				t.Fatal(err)
			}
			if !reflect.DeepEqual(tc.emptyObj, tc.obj) {
				t.Fatalf("round-tripped case-insensitive diff: %s", cmp.Diff(tc.obj, tc.emptyObj))
			}
		})
	}
}
```

**Comment by damemi** [comment]
@liggitt ah right, that makes sense. Updated

**Comment by liggitt** [comment]
/approve

lgtm, will probably need to fixup the publishing rules yaml file (the verify job will tell you if that's needed). will leave final lgtm to @kubernetes/sig-scheduling-pr-reviews

File: staging/src/k8s.io/kube-scheduler/extender/v1/types_test.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+	"github.com/google/go-cmp/cmp"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func TestCompatibility(t *testing.T) {
```
Comment (alculquicondor) [inline staging/src/k8s.io/kube-scheduler/extender/v1/types_test.go]: Add a comment indicating what is this checking and why. Perhaps also leave a TODO with an open issue for a v2.

**Review by alculquicondor — COMMENTED** [review]
LGTM after all tests are fixed :)

File: staging/src/k8s.io/kube-scheduler/extender/v1/types_test.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// TestCompatibility verifies that the types in extender/v1 can be successfully encoded to json and decoded back, even when lowercased,
+// since these types were written around JSON tags and we need to enforce consistency on them now.
+// @TODO: v2 of these types should be defined with proper JSON tags to enforce this
```
Comment (alculquicondor) [inline staging/src/k8s.io/kube-scheduler/extender/v1/types_test.go]: are you missing a sentence here? :sweat_smile:

Comment (alculquicondor) [reply staging/src/k8s.io/kube-scheduler/extender/v1/types_test.go]: Aldo, prefer something like `TODO(#1234)` for tracking

Comment (damemi) [reply staging/src/k8s.io/kube-scheduler/extender/v1/types_test.go]: I don't think so, is there something you'd like me to add? Opened an issue for this here: https://github.com/kubernetes/kubernetes/issues/88634

Comment (alculquicondor) [reply staging/src/k8s.io/kube-scheduler/extender/v1/types_test.go]: It's not clear what "this" means in `v2 of these types should be defined with proper JSON tags to enforce this`

Comment (liggitt) [reply staging/src/k8s.io/kube-scheduler/extender/v1/types_test.go]: I would just reference the issue for making the API more consistent with other serialized APIs, and put details in that issue.

Comment (damemi) [reply staging/src/k8s.io/kube-scheduler/extender/v1/types_test.go]: Did a little of both, I agree we can discuss details in the issue

**Comment by damemi** [comment]
@alculquicondor had to update import restrictions for kube-scheduler to allow k8s.io/api

**Review by Huang-Wei — COMMENTED** [review]
/lgtm

Thanks @damemi.

**Comment by Huang-Wei** [comment]
A side note: this PR is actually a leftover I didn't finish in last release, when I thought the serialized extender APIs would be better to behave in an unversioned way, while staging/... usually holds versioned APIs, so I left them in pkg/scheduler/apis.

Although it looks unusual, it seems @liggitt is ok with putting unversioned APIs in staging/..., so I'm glad to see this PR moving forward, so that extender users don't need to vendor k/k.

BTW: I have some thoughts on future refactoring on incorporating TypeMeta into this API, we can discuss the details in #88634.

**Comment by liggitt** [comment]
> Although it looks unusual, it seems @liggitt is ok with putting unversioned APIs in staging/..., so I'm glad to see this PR moving forward, so that extender users don't need to vendor k/k.

It was already exposed as an API via the extender framework. Keeping the go types in a hard-to-use location didn't improve that. Making our API types accessible, adding tests to ensure we stay compatible with them, and making plans to improve their structure in the future is the best thing we can do.

**Comment by mailhari** [comment]
How do I access this kubeschedulerconfiguration API (like I access kube-apiserver API's). I wanted to configure my extender through the API instead of  touching manifest.json. Is it possible? I couldnt find any example - about the URL format. Also I dont get any clue from scheduler code.

**Comment by alculquicondor** [comment]
The kubeschedulerconfiguration API is startup only. You have to provide a file to your kube-scheduler binary.

**Comment by mailhari** [comment]
You mean there is no other way other than providing through kube-scheduler binary?

**Comment by alculquicondor** [comment]
correct

**Pull request merged**